multibranch/src/test/java/org/jenkinsci/plugins/workflow/multibranch/SCMBinderTest.java

```diff
@@ -229,4 +231,56 @@
         });
     }
 
+    /**
+     * This test aims to verify that {@link WorkflowMultiBranchProject#CRITERIA} is being applied.
```

Comment:
This is exactly backwards. The criteria were already being applied _before_ the main patch. With that patch, `AbstractGitSCMSource.retrieve` is called, which does _not_ check branch criteria. Probably it was not meant to; `SCMSourceOwner.getSCMSourceCriteria` talks about a “candidate” head, whereas `SCMSource.retrieve(SCMHead, TaskListener)` is allowed to return null but only if it “no longer exists” (no mention of criteria).

This test _does_ reproduce one instance of JENKINS-30798, though not the one I have observed, nor the one reported in JIRA; there could be multiple root causes.